### PR TITLE
feat: unified quiet-period debounce with per-role override (fixer first)

### LINKED ISSUE #632: feat: unified quiet-period debounce with per-role override (fixer first)
## Summary

Looper should support a **shared quiet-period debounce** for role loops: when a new actionable signal arrives, do not start work immediately—wait N seconds, and **reset the wait** if another signal arrives during the window.

Today this exists only for **Reviewer** (`roles.reviewer.behavior.loop.quietPeriodSeconds`). **Fixer** starts as soon as new fixable items are discovered, which causes thrash when review comments / CI / conflicts arrive in bursts.

This issue asks for:

1. A **unified config default** for quiet period
2. **Per-role override** (and existing project-level role override)
3. **Fixer implementation first**, with Reviewer wired onto the same inheritance chain without changing its current default behavior

---

### Fixer

When discovery finds new fixable items, Fixer generally enqueues with `AvailableAt = now` and the run can start immediately.

That is painful when:

- several review comments land within seconds
- CI flips while comments are still arriving
- a reviewer is still mid-pass and will leave more threads

Ideal behavior: wait a quiet window; each new fixable-set change **resets** the timer; only start after the set has been stable for N seconds.

### Reviewer (already partial)

Reviewer already has continuous-follow-up debounce via:

- `roles.reviewer.behavior.loop.quietPeriodSeconds` (default 60)
- queued follow-up extends when a new head is seen (`TestDiscoverPullRequestsExtendsDebounceWhenQueuedFollowUpSeesNewHead`)

This should remain, but inherit from a shared default instead of being a one-off shape forever.

### What is *not* the same problem

These existing delays must stay separate knobs:

| Mechanism | Role | Purpose |
|-----------|------|---------|
| failure retry backoff | several | retry after errors |
| fixer no-op / follow-up backoff | fixer | same fix-set did not progress |
| `roles.coordinator.dispatch.autonomous.delayMinutes` | coordinator | dispatch delay, not signal-settling debounce |
| `minPublishIntervalSeconds` | reviewer | publish rate limit, not discovery debounce |

Quiet period = **settle new actionable signals before starting work**. Do not collapse retry/publish/dispatch into it.

---

## Goals

1. Shared default: `defaults.loop.quietPeriodSeconds`
2. Per-role override: `roles.<role>.behavior.loop.quietPeriodSeconds`
3. Project override via existing `projects[].roles.*` partial merge
4. Fixer honors quiet period for new / changed fixable sets, with **reset-on-new-signal**
5. Reviewer keeps current semantics and default (60), but resolves effective value through the same inheritance chain
6. Other roles (planner/worker/coordinator) are **out of scope** unless a clear signal source is defined later

## Non-goals

- Changing Coordinator autonomous dispatch delay
- Unifying retry backoff / no-op follow-up into quiet period
- Enabling quiet period for planner/worker in this issue
- Hot-reload of quiet period (keep **restart-bound**, same class as reviewer timing fields that feed durable queue `AvailableAt`)

---

## Proposed config shape

```toml
# Shared default (new)
[defaults.loop]
quietPeriodSeconds = 60   # integer >= 0; 0 = off

# Role overrides (omit => inherit defaults.loop)
[roles.reviewer.behavior.loop]
quietPeriodSeconds = 60   # already exists; keep path + default behavior

[roles.fixer.behavior.loop]
quietPeriodSeconds = 0    # new path; default 0 (opt-in) for zero surprise

# Optional project overlay (existing roles partial merge)
[[projects]]
id = "demo"
[projects.roles.fixer.behavior.loop]
quietPeriodSeconds = 120
```

### Effective value resolution

```
effective(role) =
  projects[].roles.<role>.behavior.loop.quietPeriodSeconds
  ?? roles.<role>.behavior.loop.quietPeriodSeconds
  ?? defaults.loop.quietPeriodSeconds
  ?? role-specific hardcoded default
```

Suggested hardcoded fallbacks after inheritance miss:

- Reviewer: `60` (preserve today)
- Fixer: `0` (off until configured — safest migration)

### Env / CLI

Follow existing config mapping conventions, e.g.:

- `LOOPER_DEFAULTS_LOOP_QUIET_PERIOD_SECONDS`
- `LOOPER_ROLES_FIXER_BEHAVIOR_LOOP_QUIET_PERIOD_SECONDS`
- existing reviewer env/flag paths remain valid aliases

### Why this shape (trade-off)

| Option | Why reject / accept |
|--------|---------------------|
| Global-only, no role override | Reject — reviewer vs fixer want different windows |
| Parallel top-level `debounce.roles.*` tree | Reject — forks from `roles.*.behavior.loop`, doubles docs/CLI/hot-reload surface |
| Nested under each role only, no shared default | Acceptable but weaker — operators re-copy the same number |
| **`defaults.loop` + role override** | Accept — matches merge model, reuses reviewer path |

**New concept cost:** one shared default object (`defaults.loop`) + fixer `behavior.loop` section. Cost is justified because Fixer needs the behavior and Reviewer already proved the knob; a second parallel tree would cost more long-term.

**Authority:** queue `AvailableAt` / loop `NextRunAt` remain the scheduler authority. Quiet period only decides the initial/extended eligible time when signals change — it does not invent a second scheduler.

---

### Shared primitive

Something like:

```text
DebounceSchedule(now, quiet, existingAvailableAt, mode) -> nextAvailableAt
```

- `quiet <= 0` → eligible immediately (`now`)
- on **new/changed signal** while target is still schedulable: `next = max(existing, now+quiet)` (**extend/reset**, never shorten due to debounce)
- distinguish from today's fixer enqueue helper that may **pull earlier** for same-dedupe urgency — debounce reset is the opposite direction and must be explicit

### Fixer (primary implementation)

**Signal:** fixable set identity for a PR/loop — practically `fixItemsStateHash` / unresolved thread set change (plus whatever discovery already treats as “new work”), not merely poll ticks with identical state.

**Rules:**

1. **New or changed fixable set, loop idle/queued:** enqueue/update with `AvailableAt = now + quiet` (when quiet > 0).
2. **Still queued, set changes again:** reset/extend `AvailableAt` to `now + quiet` (take later time).
3. **Identical set, no new signal:** do not keep pushing `AvailableAt` out forever on every discovery poll.
4. **Already running:** do not interrupt; keep pending rediscovery; after run ends, apply quiet period before the next start if the pending set is still actionable.
5. **Quiet vs existing backoffs:** failure retry and no-op follow-up backoff remain separate; if both apply, eligible time is the **max** of the relevant constraints (document the precedence in code/docs).
6. **Dedupe key caveat (P0 design constraint):** fixer dedupe keys include `FixItemsHash`. A hash change must **not** bypass debounce by creating a brand-new immediately-runnable queue item beside an existing delayed one. Debounce must be enforced at **loop/PR** scope (coalesce active queue work for that fixer loop), not only same-dedupe-key scope.

### Reviewer (compat)

- Keep continuous follow-up quiet-period semantics.
- Resolve config via inheritance chain above.
- No intentional behavior change when only the old reviewer field is set.
- If both `defaults.loop.quietPeriodSeconds` and reviewer override exist, reviewer override wins.

### Other roles

Out of scope for implementation in this issue. Document extension point only.

---

## Restart / hot-reload

Classify new quiet-period fields like existing reviewer timing inputs:

- **restart-bound** (durable `AvailableAt` / `NextRunAt`)
- mixed hot-safe + restart-bound edits still rejected as one candidate

Update:

- `internal/config` validation (`>= 0`)
- hot-reload classification tables / tests
- `docs/configuration.md` + skill config reference

---

### Config

- [ ] `defaults.loop.quietPeriodSeconds` exists, validated `>= 0`
- [ ] `roles.fixer.behavior.loop.quietPeriodSeconds` exists, validated `>= 0`, default `0`
- [ ] Reviewer effective quiet period inherits `defaults.loop` when reviewer field unset
- [ ] Project-level `projects[].roles.fixer.behavior.loop.quietPeriodSeconds` overrides global role value
- [ ] Env/CLI/docs/skill references updated; deprecated paths not required for the new fixer field
- [ ] Restart-bound classification covered by tests

### Fixer behavior

- [ ] With quiet `> 0`, newly discovered fixable work is **not** immediately runnable
- [ ] A second signal inside the window **resets** the wait (extends `AvailableAt`)
- [ ] After a full quiet window with no newer signal, fixer becomes eligible and can run
- [ ] Discovery polls with **unchanged** fixable set do not indefinitely postpone eligibility
- [ ] Running loop is not cancelled by debounce; post-run pending work is scheduled with quiet period
- [ ] Hash/dedupe changes cannot bypass loop-scoped debounce
- [ ] Retry / no-op follow-up backoff still work and compose safely with quiet period
- [ ] With quiet `0` (default), behavior matches today’s immediate enqueue (compat)

### Reviewer behavior

- [ ] Existing quiet-period tests still pass
- [ ] Inheritance from `defaults.loop` covered by config/normalize tests
- [ ] Explicit reviewer override still wins

### Tests (prefer contract/integration over unit-only)

- [ ] Config normalize/validate/inheritance/project overlay
- [ ] Fixer discovery debounce + reset-on-new-item
- [ ] Fixer unchanged-set poll does not extend forever
- [ ] Fixer loop-scoped coalesce when fixItemsHash changes during delay
- [ ] Fixer mid-run pending rediscovery + quiet
- [ ] Reviewer regression for existing debounce extend-on-new-head

---

### Phase 1 — Config surface

- Add `defaults.loop` + fixer `behavior.loop`
- Wire normalize / validate / env / CLI / docs
- Point reviewer effective resolution at shared inheritance without behavior change

### Phase 2 — Shared schedule helper

- Pure helper for quiet-period eligible time (extend-only on new signal)
- Unit tests for compose-with-existing-availableAt edge cases

### Phase 3 — Fixer enqueue/discovery

- Loop-scoped debounce on new/changed fixable sets
- Fix coalesce so hash changes cannot skip the window
- Contract tests listed above

### Phase 4 — Docs / operator notes

- Document defaults, overrides, restart-bound, and “quiet vs retry vs no-op backoff”
- Call out recommended starter values (e.g. fixer 60–120s) as examples only

---

### New concept trade-off

- **Failure prevented:** fixer thrash on bursty review/CI signals; duplicated one-off timing knobs per role
- **Cost:** new default object; fixer behavior.loop section; loop-scoped coalesce rules; restart-bound surface; more enqueue edge cases when hash changes mid-delay
- **Why not simpler:** “just sleep in discovery” is non-durable across restarts; “reuse reviewer field only” leaves fixer unsolved; “global-only” is too coarse

### Authority

> What is the authority for this action, and why is it not the agent's own structured output?

Scheduler eligibility remains **`AvailableAt` / `NextRunAt`** derived from discovery signals (fixable-set / head identity). Quiet period only delays eligibility after signal changes; it does not trust agent output and does not add an inference layer over infra beyond existing discovery hashes.

### Prefer deletion?

Evaluated: cannot delete reviewer quiet period (already load-bearing). Extending the same concept to fixer + a shared default is smaller than inventing a second debounce subsystem.

---

## References

- Reviewer quiet period: `roles.reviewer.behavior.loop.quietPeriodSeconds` (`internal/config`, `internal/reviewer/runner.go`)
- Reviewer extend-on-new-head tests: `TestDiscoverPullRequestsDebouncesContinuousFollowUp`, `TestDiscoverPullRequestsExtendsDebounceWhenQueuedFollowUpSeesNewHead`
- Fixer enqueue currently allows moving delayed same-dedupe items **earlier**, not debounce-extend: `internal/fixer/runner.go` (`enqueue`)
- Fixer immediate rediscovery after run uses `AvailableAt: now` today (`schedulePendingRediscoveryAfterRun`)

---

## Open choices (implementer may decide, document in PR)

1. **Fixer default `0` vs non-zero** — this issue recommends `0` for compat; product may later flip default once soak-tested.
2. **Exact fixable-set identity** for “signal changed” — should track the same hash discovery already uses for work identity; if issue comments vs threads differ, keep one authoritative hash.
3. **Whether mid-delay newer hash replaces queue payload in place vs cancel+recreate** — either is fine if loop-scoped eligibility and single active work item invariants hold.

### PR DESCRIPTION
**Pull request by mrcfps**
## Summary

Implements shared quiet-period debounce for role loops so bursty review/CI signals settle before work starts, with per-role and project overrides. Fixer is the primary consumer; Reviewer keeps current behavior while resolving config through the same inheritance chain.

- Add `defaults.loop.quietPeriodSeconds` (default 60) and `roles.fixer.behavior.loop.quietPeriodSeconds` (default 0, opt-in)
- Shared `loops.DebounceSchedule` helper (extend-only on new signal; compose with backoff via max)
- Fixer discovery applies quiet period for new/changed fixable sets, resets on signal change, does not extend forever on unchanged polls, and coalesces loop-scoped queue work so `fixItemsHash` changes cannot bypass the delay
- Reviewer inherits `defaults.loop` when its field is unset; explicit reviewer override still wins
- Restart-bound classification, env/CLI, docs, and skill config references updated

## Design notes

**Trade-off:** one shared `defaults.loop` object + fixer `behavior.loop` is cheaper than a parallel debounce tree; Fixer needs the behavior and Reviewer already proved the knob.

**Authority:** queue `AvailableAt` / loop `NextRunAt` remain scheduler authority. Quiet period only delays eligibility after discovery signal changes.

**Open choices:**
1. Fixer default stays `0` for zero-surprise migration (can flip later after soak).
2. Signal identity uses discovery `fixItemsStateHash`.
3. Mid-delay hash changes update the active queue item in place (loop-scoped coalesce).

## Test plan

- [x] `go test ./internal/config/ ./internal/fixer/ ./internal/loops/`
- [x] Reviewer debounce regression tests (`Debounce|ExtendsDebounce`)
- [x] Config inheritance / project overlay / validation / parity fixtures
- [x] Fixer: delay, reset-on-change, unchanged-set no-extend, loop coalesce, post-run quiet, quiet=0 compat
- [x] `go build ./...` and `go vet` on touched packages

Closes #632

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=worker · agent=grok-build · An autonomous AI dev team for your GitHub repos.</sub>

### CODE REVIEW

File: internal/config/load.go [hunk 1]
```diff
@@ -894,6 +894,28 @@ func parseCLIArgs(args []string) (parsedCLIArgs, error) {
 			}
 			ensureReviewerLoopConfig(&parsed.overrides).QuietPeriodSeconds = parsedValue
 			index = nextIndex
+		case matchesFlag(arg, "--defaults-loop-quiet-period-seconds"):
+			value, nextIndex, err := takeValue(index, "--defaults-loop-quiet-period-seconds")
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply internal/config/load.go:898]: @chatgpt-codex-connector This has been fixed in 88688dd.

Registered --defaults-loop-quiet-period-seconds and --roles-fixer-behavior-loop-quiet-period-seconds in globalFlagSpecs and configFlagNames in internal/cliapp/app.go so Cobra accepts them and ExtractConfigArgs forwards them. Updated root/daemon help goldens and added forwarding coverage.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=grok-build · An autonomous AI dev team for your GitHub repos.</sub>

File: internal/cliapp/config_commands.go [hunk 2]
```diff
@@ -37,9 +37,19 @@ var configFieldRegistry = map[string]configField{
...
-	"package.autoUpgradeEnabled":  boolFieldWithAlias("package.autoUpgradeEnabled", "LOOPER_AUTO_UPGRADE_ENABLED", "", "package-auto-upgrade-enabled", "no-auto-upgrade", func(c config.Config) any { return c.Package.AutoUpgradeEnabled }, func(p *config.PartialConfig) **bool { return &ensurePartialPackage(p).AutoUpgradeEnabled }),
-	"instructions.maxBytes":       positiveIntField("instructions.maxBytes", "", "", func(c config.Config) any { return c.Instructions.MaxBytes }, func(p *config.PartialConfig) **int { return &ensurePartialInstructions(p).MaxBytes }),
+	"defaults.loop.quietPeriodSeconds": nonNegativeIntField("defaults.loop.quietPeriodSeconds", "LOOPER_DEFAULTS_LOOP_QUIET_PERIOD_SECONDS", "defaults-loop-quiet-period-seconds", func(c config.Config) any {
+		return c.Defaults.Loop.QuietPeriodSeconds
+	}, func(p *config.PartialConfig) **int {
+		return &ensurePartialDefaultsLoop(p).QuietPeriodSeconds
+	}),
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply internal/cliapp/config_commands.go:44]: @chatgpt-codex-connector This has been fixed in 88688dd.

Added configFieldSet cases for defaults.loop.quietPeriodSeconds and roles.fixer.behavior.loop.quietPeriodSeconds in internal/cliapp/config_commands.go so config show --source reports config-file or cli instead of default. Covered both paths in a regression test.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=grok-build · An autonomous AI dev team for your GitHub repos.</sub>

File: internal/config/types.go [hunk 3]
```diff
@@ -393,16 +393,23 @@ type ProjectNetworkConfig struct {
...
+	BaseBranch         string             `json:"baseBranch"`
+	AllowAutoCommit    bool               `json:"allowAutoCommit"`
+	AllowAutoPush      bool               `json:"allowAutoPush"`
+	AllowAutoApprove   bool               `json:"allowAutoApprove"`
+	AllowAutoMerge     bool               `json:"allowAutoMerge"`
+	AllowRiskyFixes    bool               `json:"allowRiskyFixes"`
+	FixAllPullRequests bool               `json:"fixAllPullRequests"`
+	OpenPRStrategy     OpenPRStrategy     `json:"openPrStrategy"`
+	AddSnapshotMode    AddSnapshotMode    `json:"addSnapshotMode"`
+	Loop               DefaultsLoopConfig `json:"loop"`
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply internal/config/types.go:415]: @chatgpt-codex-connector This has been fixed in 88688dd.

Updated internal/api/testdata/contracts/daemon-http.responses.compat.json config.get body with defaults.loop.quietPeriodSeconds (60) and roles.fixer.behavior.loop.quietPeriodSeconds (0). config.patch reuses that body via sameAs; core route artifact test now passes.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=grok-build · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by mrcfps** [comment]
**Looper fixer round complete** — 88688dd

- ✅ Review comment on `internal/config/load.go:898` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/633#discussion_r3710097078)
  - Registered --defaults-loop-quiet-period-seconds and --roles-fixer-behavior-loop-quiet-period-seconds in globalFlagSpecs and configFlagNames in internal/cliapp/app.go so Cobra accepts them and ExtractConfigArgs forwards them. Updated root/daemon help goldens and added forwarding coverage.
- ✅ Review comment on `internal/cliapp/config_commands.go:44` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/633#discussion_r3710097085)
  - Added configFieldSet cases for defaults.loop.quietPeriodSeconds and roles.fixer.behavior.loop.quietPeriodSeconds in internal/cliapp/config_commands.go so config show --source reports config-file or cli instead of default. Covered both paths in a regression test.
- ✅ Review comment on `internal/config/types.go:412` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/633#discussion_r3710097091)
  - Updated internal/api/testdata/contracts/daemon-http.responses.compat.json config.get body with defaults.loop.quietPeriodSeconds (60) and roles.fixer.behavior.loop.quietPeriodSeconds (0). config.patch reuses that body via sameAs; core route artifact test now passes.
- • Review comment on `internal/config/validate.go:229` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/633#discussion_r3710151416)
  - Agent did not provide a decision for this thread


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=grok-build · An autonomous AI dev team for your GitHub repos.</sub>

File: internal/config/validate.go [hunk 4]
```diff
@@ -218,9 +218,15 @@ func ValidateWithOptions(config Config, options ValidateOptions) error {
...
 
+	if config.Defaults.Loop.QuietPeriodSeconds < 0 {
+		issues = append(issues, ValidationIssue{Path: "defaults.loop.quietPeriodSeconds", Message: "must be an integer >= 0"})
+	}
 	if config.Roles.Reviewer.Behavior.Loop.QuietPeriodSeconds < 0 {
 		issues = append(issues, ValidationIssue{Path: "roles.reviewer.behavior.loop.quietPeriodSeconds", Message: "must be an integer >= 0"})
 	}
+	if config.Roles.Fixer.Behavior.Loop.QuietPeriodSeconds < 0 {
+		issues = append(issues, ValidationIssue{Path: "roles.fixer.behavior.loop.quietPeriodSeconds", Message: "must be an integer >= 0"})
+	}
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply internal/config/validate.go]: @chatgpt-codex-connector This has been fixed in bf4cb92.

validateProjectRoleOverrides now rejects negative projects[].roles.fixer/reviewer.behavior.loop.quietPeriodSeconds so invalid project overlays fail fast. Added TestQuietPeriodValidationRejectsNegativeProjectOverride covering both paths.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=grok-build · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by mrcfps** [comment]
**Looper fixer round complete** — bf4cb92

- ✅ Review comment on `internal/config/validate.go:229` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/633#discussion_r3710151416)
  - validateProjectRoleOverrides now rejects negative projects[].roles.fixer/reviewer.behavior.loop.quietPeriodSeconds so invalid project overlays fail fast. Added TestQuietPeriodValidationRejectsNegativeProjectOverride covering both paths.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=grok-build · An autonomous AI dev team for your GitHub repos.</sub>

File: internal/config/types.go [hunk 5]
```diff
@@ -393,16 +393,23 @@ type ProjectNetworkConfig struct {
 	Mode NetworkMode `json:"mode,omitempty"`
 }
 
+type DefaultsLoopConfig struct {
+	// QuietPeriodSeconds is the shared default settle window for role loops.
+	// 0 disables quiet-period debounce. Role overrides win when set.
+	QuietPeriodSeconds int `json:"quietPeriodSeconds"`
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply internal/config/types.go:402]: @chatgpt-codex-connector This has been fixed in 0a0e7d0.

Documented the full quiet-period debounce trade-off in docs/configuration.md under Concept trade-off (failure prevented; costs for shared config, fixItemsStateHash, queue coalesce, edge cases, and failure modes; why deleting the layer, trusting agent output, or failing loud is insufficient), and linked it from DefaultsLoopConfig in internal/config/types.go.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=grok-build · An autonomous AI dev team for your GitHub repos.</sub>

File: internal/loops/quiet_period.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+//   - otherwise next = max(existingAvailableAt, now+quiet)
+//
+// Existing AvailableAt is never shortened by debounce (extend/reset only).
+// Retry backoff and no-op follow-up delays remain separate constraints; callers
+// that need both should take the max of DebounceSchedule and those times.
+func DebounceSchedule(now time.Time, quietSeconds int, existingAvailableAt time.Time) time.Time {
+	if quietSeconds <= 0 {
+		return now
+	}
+	candidate := now.Add(time.Duration(quietSeconds) * time.Second)
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply internal/loops/quiet_period.go:18]: @chatgpt-codex-connector This has been fixed in 0a0e7d0.

Added validateQuietPeriodSeconds so global and project quietPeriodSeconds values that overflow time.Duration are rejected at config validation (same bound as agent timeouts), with regression tests in internal/config/quiet_period_test.go.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=grok-build · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by mrcfps** [comment]
**Looper fixer round complete** — 0a0e7d0

- ✅ Review comment on `internal/config/types.go:402` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/633#discussion_r3710243828)
  - Documented the full quiet-period debounce trade-off in docs/configuration.md under Concept trade-off (failure prevented; costs for shared config, fixItemsStateHash, queue coalesce, edge cases, and failure modes; why deleting the layer, trusting agent output, or failing loud is insufficient), and linked it from DefaultsLoopConfig in internal/config/types.go.
- ✅ Review comment on `internal/loops/quiet_period.go:18` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/633#discussion_r3710243832)
  - Added validateQuietPeriodSeconds so global and project quietPeriodSeconds values that overflow time.Duration are rejected at config validation (same bound as agent timeouts), with regression tests in internal/config/quiet_period_test.go.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=grok-build · An autonomous AI dev team for your GitHub repos.</sub>

File: internal/fixer/runner.go [hunk 7]
```diff
@@ -5050,24 +5064,52 @@ func (r *Runner) ensureLoopForPullRequest(ctx context.Context, project storage.P
...
+		if err != nil {
+			return loopUpsertResult{}, err
+		}
+		existingAvailableAt := time.Time{}
+		signalChanged := true
+		if activeQueue != nil && activeQueue.Status == "queued" {
+			existingAvailableAt = parseRFC3339OrZero(activeQueue.AvailableAt)
+			queuedHash := fixItemsHashFromQueueItem(*activeQueue)
+			// Identical fixable-set polls must not push AvailableAt out forever.
+			signalChanged = queuedHash == "" || queuedHash != strings.TrimSpace(fixItemsStateHash)
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply internal/fixer/runner.go]: @chatgpt-codex-connector This has been fixed in 470e8d4.

signalChanged now compares the queued head SHA as well as fixItemsStateHash in ensureLoopForPullRequest (internal/fixer/runner.go). A new commit with unchanged fix items resets the quiet window and replaces the loop-scoped queue payload. Added TestDiscoverHeadSHAChangeResetsQuietWindow.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=grok-build · An autonomous AI dev team for your GitHub repos.</sub>

File: internal/cliapp/config_commands.go [hunk 8]
```diff
@@ -1578,6 +1625,10 @@ func configFieldSet(partial config.PartialConfig, key string) bool {
 			return false
 		}
 		return partial.Defaults.OpenPRStrategy != nil
+	case "defaults.loop.quietPeriodSeconds":
+		return partial.Defaults != nil && partial.Defaults.Loop != nil && partial.Defaults.Loop.QuietPeriodSeconds != nil
+	case "roles.fixer.behavior.loop.quietPeriodSeconds":
+		return partial.Roles != nil && partial.Roles.Fixer != nil && partial.Roles.Fixer.Behavior != nil && partial.Roles.Fixer.Behavior.Loop != nil && partial.Roles.Fixer.Behavior.Loop.QuietPeriodSeconds != nil
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply internal/cliapp/config_commands.go:1654]: @chatgpt-codex-connector This has been fixed in 470e8d4.

config show --source now attributes roles.fixer.behavior.loop.quietPeriodSeconds to defaults.loop when the fixer field is not directly set by file/env/CLI (resolveConfigFieldSource in internal/cliapp/config_commands.go). Added TestConfigShowSourceAttributesInheritedFixerQuietPeriod.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=grok-build · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by mrcfps** [comment]
**Looper fixer round complete** — 470e8d4

- ✅ Review comment on `internal/fixer/runner.go` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/633#discussion_r3710304354)
  - signalChanged now compares the queued head SHA as well as fixItemsStateHash in ensureLoopForPullRequest (internal/fixer/runner.go). A new commit with unchanged fix items resets the quiet window and replaces the loop-scoped queue payload. Added TestDiscoverHeadSHAChangeResetsQuietWindow.
- ✅ Review comment on `internal/cliapp/config_commands.go:1654` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/633#discussion_r3710304356)
  - config show --source now attributes roles.fixer.behavior.loop.quietPeriodSeconds to defaults.loop when the fixer field is not directly set by file/env/CLI (resolveConfigFieldSource in internal/cliapp/config_commands.go). Added TestConfigShowSourceAttributesInheritedFixerQuietPeriod.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=grok-build · An autonomous AI dev team for your GitHub repos.</sub>
